internal/datamodel/confirm_project_commitments.go

```diff
@@ -171,43 +280,106 @@ func ConfirmPendingCommitments(loc core.AZResourceLocation, cluster *core.Cluste
+	confirmedCommitmentIDs []db.ProjectCommitmentID, transferredCommitmentIDs map[db.ProjectCommitmentID]core.CommitmentTransferLeftover, now time.Time) (result []db.MailNotification, err error) {
+
 	var n core.CommitmentGroupNotification
-	err := dbi.QueryRow("SELECT d.name, p.name FROM domains d JOIN projects p ON d.id = p.domain_id where p.id = $1", projectID).Scan(&n.DomainName, &n.ProjectName)
+	err = dbi.QueryRow("SELECT d.name, p.name FROM domains d JOIN projects p ON d.id = p.domain_id where p.id = $1", projectID).Scan(&n.DomainName, &n.ProjectName)
 	if err != nil {
-		return db.MailNotification{}, err
+		return result, err
 	}
 
-	var commitments []db.ProjectCommitment
-	_, err = dbi.Select(&commitments, `SELECT * FROM project_commitments WHERE id = ANY($1)`, pq.Array(confirmedCommitmentIDs))
+	commitmentsByID, err := db.BuildIndexOfDBResult(dbi, func(c db.ProjectCommitment) db.ProjectCommitmentID { return c.ID }, `SELECT * FROM project_commitments WHERE id = ANY($1)`, pq.Array(append(confirmedCommitmentIDs, slices.Collect(maps.Keys(transferredCommitmentIDs))...)))
 	if err != nil {
-		return db.MailNotification{}, err
+		return result, err
 	}
-	for _, c := range commitments {
+	for _, cID := range confirmedCommitmentIDs {
+		c, exists := commitmentsByID[cID]
+		if !exists {
+			return result, fmt.Errorf("tried to generate mail notification for non-existent commitment ID %d", cID)
+		}
 		confirmedAt := c.ConfirmedAt.UnwrapOr(time.Unix(0, 0)) // the UnwrapOr() is defense in depth, it should never be relevant because we only notify for confirmed commitments here
 		n.Commitments = append(n.Commitments, core.CommitmentNotification{
 			Commitment: c,
 			DateString: confirmedAt.Format(time.DateOnly),
 			Resource:   loc,
 		})
 	}
+	if len(n.Commitments) != 0 {
+		newNotification, err := tplConfig.ConfirmedCommitments.Render(n, projectID, now)
+		if err != nil {
+			return result, err
+		}
+		result = append(result, newNotification)
+	}
 
-	return tpl.Render(n, projectID, now)
+	n.Commitments = make([]core.CommitmentNotification, 0, len(transferredCommitmentIDs))
+	for cID, leftover := range transferredCommitmentIDs {
+		c, exists := commitmentsByID[cID]
+		if !exists {
+			return result, fmt.Errorf("tried to generate mail notification for non-existent commitment ID %d", cID)
+		}
+		confirmedAt := c.ConfirmedAt.UnwrapOr(time.Unix(0, 0)) // the UnwrapOr() is defense in depth, it should never be relevant because we only notify for confirmed commitments here
+		n.Commitments = append(n.Commitments, core.CommitmentNotification{
+			Commitment: c,
+			DateString: confirmedAt.Format(time.DateOnly),
+			Resource:   loc,
+			Leftover:   leftover,
+		})
+	}
+	if len(n.Commitments) == 0 {
+		return result, nil
+	}
+	newNotification, err := tplConfig.TransferredCommitments.Render(n, projectID, now)
+	if err != nil {
+		return result, err
+	}
+	result = append(result, newNotification)
```

Comment:
I know (and agree with) the notion of returning early, but in this case, I got confused because this looked different from L354-L360 above where the other notification is rendered. Since it is the same, I would like for it to look the same:
```suggestion
	if len(n.Commitments) != 0 {
		newNotification, err := tplConfig.TransferredCommitments.Render(n, projectID, now)
		if err != nil {
			return result, err
		}
		result = append(result, newNotification)
	}
```